### DISCUSSION TITLE
Push Branch to private repo

### DISCUSSION THREAD

**Discussion by kidCaulfield** [community/community#26471] [Actions]
<p>Hi,</p>
<p>I’m trying to build a workflow that will push a staging branch to a remote repo on a staging server. However the action results in this error:</p>
<p>=========================================================</p>
<p>! [rejected] staging -&gt; staging (fetch first)</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:17" rel="nofollow noopener">17</a>error: failed to push some refs to ‘privat@repo.git’</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:18" rel="nofollow noopener">18</a>hint: Updates were rejected because the remote contains work that you do</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:19" rel="nofollow noopener">19</a>hint: not have locally. This is usually caused by another repository pushing</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:20" rel="nofollow noopener">20</a>hint: to the same ref. You may want to first integrate the remote changes</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:21" rel="nofollow noopener">21</a>hint: (e.g., ‘git pull …’) before pushing again.</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:22" rel="nofollow noopener">22</a>hint: See the ‘Note about fast-forwards’ in ‘git push --help’ for details.</p>
<p><a href="https://github.com/erant22/onecomply-api/runs/613287055?check_suite_focus=true#step:3:23" rel="nofollow noopener">23</a>##[error]Process completed with exit code 1.</p>
<p>==========================================================</p>
<p>I’m having a had time understanding how this is possible considering that im merge my lastest commits from master into staging. Thus its should be ahead of the private remote repo on my staging server.</p>
<p>Staging Deployment workflow:</p>
<pre><code>name: CI
on:
  push:
    branches:
     - staging

jobs:
  deploy-staging:
    runs-on: ubuntu-latest
    env:
       SSH_KEY: ${{ secrets.SSH_KEY }}
    steps:
    - uses: actions/checkout@v2
    - name: Deploy api to staging
      env:
        GIT_SSH_COMMAND: "ssh -o StrictHostKeyChecking=no"
        REMOTE_REPO: ${{ secrets.REPO }}
      run: |
        echo Deploying API to Staging
        eval `ssh-agent -s`
        ssh-add - &lt;&lt;&lt; "${SSH_KEY}"
        git remote add staging "${REMOTE_REPO}"
        git push staging
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<p>Check if the commit you’re trying to push is really a direct descendant of the one on the staging branch. There are a lot of operation that might mess with that, including:</p>
<ul>
<li>Merging from master with a merge request instead of fast-forward.</li>
<li>And sort of rebase on the shared history, including rebase-merge via GitHub’s web interface.</li>
</ul>

**Comment by kidCaulfield** [comment]
<p>I put a <code>git log</code> in my workflow and found that the commit SHA hash was different in the action console than in my local terminal. So that is one problem discovered.</p>

**Comment by kidCaulfield** [comment]
<p>So i think that this might be a hopeless operation. As the hashes on githubs staging branch are different than the hash from my local machine. which are the same on my staging server. I may have to come up with a different plan of attack for this.</p>

**Comment by airtower-luna** [comment]
<p>How are you creating the staging branch on GitHub? If you push from your local machine it should be exactly the same.</p>

**Comment by kidCaulfield** [comment]
<p>I have been pushing from my local machine. But I wanted to setup a deployment workflow that would trigger by merging a pull request from master into staging on github. So this is where it change.</p>

**Comment by kidCaulfield** [comment]
<p>Hey just thought I’d share this with you since you helped me firgure this out. I got the workflow working with what I intended do from the start. MIght change this deployment strategy in the future but this workes well. Here is the final solution to this problem.</p>
<pre><code>name: CI-Deploy-Staging
on:
  push:
    branches:
     - staging

jobs:
  deploy-staging:
    runs-on: ubuntu-latest
    env:
       SSH_KEY: ${{ secrets.SSH_KEY }}
    steps:
    - uses: actions/checkout@v2
    - name: Deploy api to staging
      env:
        GIT_SSH_COMMAND: "ssh -o StrictHostKeyChecking=no"
        REMOTE_REPO: ${{ secrets.STAGING_REPO }}
      run: |
        echo Deploying API to Staging
        eval `ssh-agent -s`
        ssh-add - &lt;&lt;&lt; "${SSH_KEY}"
        git remote rm origin
        git remote add origin git@github:private.git
        git fetch --unshallow origin
        echo ==============last 10 commits==================
        git log -n 10
        echo ===============================================
        git remote add staging "${REMOTE_REPO}"
        git push staging -f
</code></pre>
